ast/edit/match.go

```diff
@@ -0,0 +1,436 @@
+	case *ast.ExpressionStatement:
+		return matchExpressionStatement(n, node.(*ast.ExpressionStatement))
+	case *ast.ReturnStatement:
+		return matchReturnStatement(n, node.(*ast.ReturnStatement))
+	case *ast.VariableAssignment:
+		return matchVariableAssignment(n, node.(*ast.VariableAssignment))
+	case *ast.MemberAssignment:
+		return matchMemberAssignment(n, node.(*ast.MemberAssignment))
+	case *ast.CallExpression:
+		return matchCallExpression(n, node.(*ast.CallExpression))
+	case *ast.PipeExpression:
+		return matchPipeExpression(n, node.(*ast.PipeExpression))
+	case *ast.MemberExpression:
+		return matchMemberExpression(n, node.(*ast.MemberExpression))
+	case *ast.IndexExpression:
+		return matchIndexExpression(n, node.(*ast.IndexExpression))
+	case *ast.BinaryExpression:
+		return matchBinaryExpression(n, node.(*ast.BinaryExpression))
+	case *ast.UnaryExpression:
+		return matchUnaryExpression(n, node.(*ast.UnaryExpression))
+	case *ast.LogicalExpression:
+		return matchLogicalExpression(n, node.(*ast.LogicalExpression))
+	case *ast.ObjectExpression:
+		return matchObjectExpression(n, node.(*ast.ObjectExpression))
+	case *ast.ConditionalExpression:
+		return matchConditionalExpression(n, node.(*ast.ConditionalExpression))
+	case *ast.ArrayExpression:
+		return matchArrayExpression(n, node.(*ast.ArrayExpression))
+	case *ast.Identifier:
+		return matchIdentifier(n, node.(*ast.Identifier))
+	case *ast.PipeLiteral:
+		return matchPipeLiteral(n, node.(*ast.PipeLiteral))
+	case *ast.StringLiteral:
+		return matchStringLiteral(n, node.(*ast.StringLiteral))
+	case *ast.BooleanLiteral:
+		return matchBooleanLiteral(n, node.(*ast.BooleanLiteral))
+	case *ast.FloatLiteral:
+		return matchFloatLiteral(n, node.(*ast.FloatLiteral))
+	case *ast.IntegerLiteral:
+		return matchIntegerLiteral(n, node.(*ast.IntegerLiteral))
+	case *ast.UnsignedIntegerLiteral:
+		return matchUnsignedIntegerLiteral(n, node.(*ast.UnsignedIntegerLiteral))
+	case *ast.RegexpLiteral:
+		return matchRegexpLiteral(n, node.(*ast.RegexpLiteral))
+	case *ast.DurationLiteral:
+		return matchDurationLiteral(n, node.(*ast.DurationLiteral))
+	case *ast.DateTimeLiteral:
+		return matchDateTimeLiteral(n, node.(*ast.DateTimeLiteral))
+	case *ast.FunctionExpression:
+		return matchFunctionExpression(n, node.(*ast.FunctionExpression))
+	case *ast.Property:
+		return matchProperty(n, node.(*ast.Property))
+	default:
+		// If we were able not to find the type, than this switch is wrong
+		panic(fmt.Errorf("unknown type %q", n.Type()))
+	}
+}
+
+func matchString(s, os string) bool {
+	if len(s) > 0 && s != os {
+		return false
+	}
+	return true
+}
+
+// negative operators are ignored (invalid value for enumeration)
+func matchOperator(o, oo ast.OperatorKind) bool {
+	if o >= 0 && o != oo {
+		return false
+	}
+	return true
+}
+
+func matchLogicalOperator(o, oo ast.LogicalOperatorKind) bool {
+	if o >= 0 && o != oo {
+		return false
+	}
+	return true
+}
+
```

Comment:
Can you add a comment here describing the behavior of this function?  Is it okay for there to be extra files in `op` that do not exist in `p`?

Reply:
Fixed, files must be match